include/Dialect/TensorExt/IR/TensorExtCanonicalization.td

```diff
@@ -0,0 +1,101 @@
+#ifndef INCLUDE_DIALECT_TENSOREXT_IR_TENSOREXTCANONICALIZATION_TD_
+#define INCLUDE_DIALECT_TENSOREXT_IR_TENSOREXTCANONICALIZATION_TD_
+
+include "TensorExtOps.td"
+include "mlir/Dialect/Arith/IR/ArithOps.td"
+include "mlir/Dialect/Tensor/IR/TensorOps.td"
+include "mlir/IR/PatternBase.td"
+
+// TODO(#515): refactor these helpers to a common file with InsertRotate.td
+defvar DefOverflow = ConstantEnumCase<Arith_IntegerOverflowAttr, "none">;
+
+def MakeSingleResultVariadic: NativeCodeCall<"{ $0 }">;
+
+def IndexType : NativeCodeCall<"$_builder.getIndexType()">;
+
+def MakeMatchingTypedZero : NativeCodeCall<
+      "$_builder.getIntegerAttr("
+        "$_builder.getIntegerType("
+          "$0.getType().getIntOrFloatBitWidth()"
+        "), 0)">;
+
+def CreateIndexCastOp : NativeCodeCall<
+      "$_builder.create<arith::IndexCastOp>($0.getLoc(), $1.getType(), $0)">;
+
+def IsZero :
+    Constraint<
+      CPred<"llvm::cast<mlir::IntegerAttr>($0).getValue().isZero()">>;
+
+def AreOpposites :
```

Comment:
This and other helpers (IndexType, MakeMatchingTypedZero) seem unused - maybe they were needed for follow-ups?

(Mostly I was curious where`AreOpposites` is used - since the constraint seems more like `AreEqual`)

Reply:
Fixed. 

I replaced a pattern that combined rotate x & rotate -x with the input tensor, which was generalized to the addition pattern followed by dropping zero rotation. Just didn't remove the helpers.